hydra/grammar/OverrideParser.g4

```diff
@@ -30,31 +36,34 @@ element:
     | function
 ;
 
-argName: ID EQUAL;
-function: ID POPEN (argName? element (COMMA argName? element )* )? PCLOSE;
-
 simpleChoiceSweep:
       element (COMMA element)+                   // value1,value2,value3
 ;
 
+// Functions.
+
+argName: ID EQUAL;
+function: ID POPEN (argName? element (COMMA argName? element )* )? PCLOSE;
+
+// Data structures.
+
+listValue: BRACKET_OPEN sequence? BRACKET_CLOSE;                          // [], [1,2,3], [a,b,[1,2]]
+dictValue: BRACE_OPEN (keyValuePair (COMMA keyValuePair)*)? BRACE_CLOSE;  // {}, {a:10,b:20}
+keyValuePair: ID COLON element;
+sequence: element (COMMA element)*;
+
+// Primitive types.
+
 primitive:
       QUOTED_VALUE                               // 'hello world', "hello world"
     | (   ID                                     // foo_10
         | NULL                                   // null, NULL
         | INT                                    // 0, 10, -20, 1_000_000
         | FLOAT                                  // 3.14, -20.0, 1e-1, -10e3
         | BOOL                                   // true, TrUe, false, False
-        | DOT_PATH                               // foo.bar
         | INTERPOLATION                          // ${foo.bar}, ${env:USER,me}
-        | UNQUOTED_CHAR                          // /, -, \, +, ., $, *
+        | UNQUOTED_CHAR                          // /, -, \, +, ., $, %, *
         | COLON                                  // :
+        | ESC                                    // \\, \ , \\t, \,
```

Comment:
Adding support for escaped characters makes it consistent with OmegaConf (and will be needed for proper parsing of interpolations)

Reply:
Currently this is documented only in OmegaConf, in https://omegaconf.readthedocs.io/en/latest/usage.html#custom-interpolations: _You can use literal commas and spaces anywhere by escaping (\, and \ )_

Reply:
Why do we need this capability in Hydra?

Reply:
It's not super useful for Hydra (it gives the ability to have leading / trailing whitespaces and commas in unquoted values, but one could use quoted values as well, so it doesn't really count as a new functionality).

The main goal is to make Hydra consistent with OmegaConf. This has two advantages:
1. Being able to re-use the same grammar once we support nested interpolations (=> not having to maintain two different grammars); this is really the original motivation for this change
2. More consistent user experience

Reply:
Btw Hydra doc added in f7ccd1e

Reply:
And are escaped actually needed by in OmegaConf or are they just there to maintain backward compatibility with the initial (stoneage) implementation or interpolations?

Reply:
They could be removed by forcing people to use quoted strings instead. This would be a breaking change though. Note also that quoted strings can get quite ugly with nested interpolations (as we discussed at some point in the past), which may be a reason to try and keep supporting as many characters as possible in unquoted strings.

Reply:
I think in general quoting and escaping serves the same need.
Depending on the circumstances, one may be cleaner than the other.
supporting just space, tab and comma seems arbitrary from the perspective of Hydra: Why those characters and not 
'[]{})()=' ?
(and anything else I forgot which is a special token?)

If we are adding escaping support, we might as well make it complete.

Reply:
I'm fine with that, it's very easy to add (I think the only one you're missing is `:`)

Reply:
I went ahead and added them in 8d65ff1a

Reply:
> Okay, please add it and update the docs and the news fragment.

I did (see previous comment). Just realized I had forgotten to update the parser / doc so I made another commit (5513fdd9f) for that. I'll refactor this PR to squash a few commits together once it's ready to be merged.